cloud/azure/azure_deployment.py

```diff
@@ -0,0 +1,620 @@
+            status_message=op.properties.status_message,
+            target_resource = dict(
+                id=op.properties.target_resource.id,
+                resource_name=op.properties.target_resource.resource_name,
+                resource_type=op.properties.target_resource.resource_type
+            ) if op.properties.target_resource else None,
+            provisioning_state=op.properties.provisioning_state,
+        )
+        for op in get_failed_nested_operations(client, resource_group, operations)
+    ]
+
+def deploy_template(module, client, conn_info):
+    """
+    Deploy the targeted template and parameters
+    :param module: Ansible module containing the validated configuration for the deployment template
+    :param client: resource management client for azure
+    :param conn_info: connection info needed
+    :return:
+    """
+
+    deployment_name = conn_info["deployment_name"]
+    group_name = conn_info["resource_group_name"]
+
+    deploy_parameter = DeploymentProperties()
+    deploy_parameter.mode = module.params.get('deployment_mode')
+
+    if module.params.get('parameters_link') is None:
+        deploy_parameter.parameters = module.params.get('parameters')
+    else:
+        parameters_link = ParametersLink(
+            uri = module.params.get('parameters_link')
+        )
+        deploy_parameter.parameters_link = parameters_link
+
+    if module.params.get('template_link') is None:
+        deploy_parameter.template = module.params.get('template')
+    else:
+        template_link = TemplateLink(
+            uri = module.params.get('template_link')
+        )
+        deploy_parameter.template_link = template_link
+
+    params = ResourceGroup(location=module.params.get('location'), tags=module.params.get('tags'))
+    try:
+        client.resource_groups.create_or_update(group_name, params)
+        result = client.deployments.create_or_update(group_name, deployment_name, deploy_parameter)
+        deployment_result = result.result() # Blocking wait, return the Deployment object
+        if module.params.get('wait_for_deployment_completion'):
+            while not deployment_result.properties.provisioning_state in ['Canceled', 'Failed', 'Deleted', 'Succeeded']:
+                deployment_result = client.deployments.get(group_name, deployment_name)
+                time.sleep(module.params.get('wait_for_deployment_polling_period'))
+
+        if deployment_result.properties.provisioning_state == 'Succeeded':
+            return deployment_result
+
+        failed_deployment_operations = get_failed_deployment_operations(module, client, group_name, deployment_name)
+        module.fail_json(msg='Deployment failed. Deployment id: %s' % (deployment_result.id), failed_deployment_operations=failed_deployment_operations)
+    except CloudError as e:
+        failed_deployment_operations = get_failed_deployment_operations(module, client, group_name, deployment_name)
```

Comment:
If there is a failure of the deployment (params or template is wrong) there will be no deployment created. A 400 will be generated from line 466, then line 479 will fail with a 404.

I'd expect the 400 failure to provide failure messaging with the reason the 400 was encountered.